File: python/gentrail/export_filter.py
"""Only spans Gentrail ingests leave the process: an app's HTTP and database
spans can carry PII in attributes redaction never inspects, so they are dropped
on every Gentrail export path. The lists match spec/spans.json export_filter."""

from __future__ import annotations

from typing import Any, Iterable, List

GENAI_SIGNAL_ATTRIBUTE_PREFIXES = ("gen_ai.", "ai.", "llm.", "openinference.", "aigentrail.")
GENAI_SIGNAL_ATTRIBUTE_KEYS = frozenset({"session.id", "agent.name", "tool.name"})


def carries_genai_signal(attribute_keys: Iterable[str]) -> bool:
    return any(
        key in GENAI_SIGNAL_ATTRIBUTE_KEYS or key.startswith(GENAI_SIGNAL_ATTRIBUTE_PREFIXES)
        for key in attribute_keys
    )


class GenAISignalExporter:
    """Wraps a SpanExporter, forwarding only spans that carry a GenAI signal."""

    def __init__(self, wrapped: Any) -> None:
        self._wrapped = wrapped

    def export(self, spans: Any) -> Any:
        kept: List[Any] = [span for span in spans if carries_genai_signal((span.attributes or {}).keys())]
        if not kept:
            from opentelemetry.sdk.trace.export import SpanExportResult

            return SpanExportResult.SUCCESS
        return self._wrapped.export(kept)

    def shutdown(self) -> None:
        self._wrapped.shutdown()

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return self._wrapped.force_flush(timeout_millis)

File: go/export_filter.go
package gentrail

import (
	"context"
	"strings"

	"go.opentelemetry.io/otel/attribute"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

var (
	genAISignalAttributePrefixes = []string{"gen_ai.", "ai.", "llm.", "openinference.", "aigentrail."}
	genAISignalAttributeKeys     = map[string]bool{"session.id": true, "agent.name": true, "tool.name": true}
)

func carriesGenAISignal(attributes []attribute.KeyValue) bool {
	for _, kv := range attributes {
		key := string(kv.Key)
		if genAISignalAttributeKeys[key] {
			return true
		}
		for _, prefix := range genAISignalAttributePrefixes {
			if strings.HasPrefix(key, prefix) {
				return true
			}
		}
	}
	return false
}

// genAISignalExporter drops spans Gentrail does not ingest, so an app's HTTP
// and database spans, whose attributes redaction never inspects, never leave
// the process.
type genAISignalExporter struct {
	sdktrace.SpanExporter
}

func (e genAISignalExporter) ExportSpans(ctx context.Context, spans []sdktrace.ReadOnlySpan) error {
	kept := make([]sdktrace.ReadOnlySpan, 0, len(spans))
	for _, span := range spans {
		if carriesGenAISignal(span.Attributes()) {
			kept = append(kept, span)
		}
	}
	if len(kept) == 0 {
		return nil
	}
	return e.SpanExporter.ExportSpans(ctx, kept)
}
